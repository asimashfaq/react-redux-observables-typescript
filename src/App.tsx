import React, { useState } from 'react';
import { makeStyles, withStyles, createStyles, Theme } from '@material-ui/core/styles';
import './App.css';
import { TextField, Paper, Button, FormControl } from '@material-ui/core';
import Typography from '@material-ui/core/Typography';
import logo from './resources/platter.svg';
import { useHistory} from "react-router-dom";
import { ISystemState } from './state-mgmt/store/system/types';

const ValidationTextField = withStyles({
  root: {
    '& input:valid + fieldset': {
      borderColor: 'green',
      borderWidth: 2,
    },
    '& input:invalid + fieldset': {
      borderColor: 'red',
      borderWidth: 2,
    },
    '& input:valid:focus + fieldset': {
      borderLeftWidth: 6,
      padding: '4px !important', // override inline-style
    },
  },
})(TextField);

interface OwnProps {
  doLogin: Function,
  login: ISystemState
}

// Main functional component login page
const App: React.FC<OwnProps> = (props: OwnProps) => {
  const [username, setUsername] = useState("");
  const [password, setPassword] = useState("");
  // High level CSS styling for the Login page. Can be refactored to another file
  // We should use CSS Grid https://css-tricks.com/snippets/css/complete-guide-grid/
  const useStyles = () => {
    return makeStyles((theme: Theme) =>
      createStyles({
        container: {
          display: 'grid',
          gridTemplateColumns: '20em auto 25em auto 20em',
          gridTemplateRows: '5em 25em auto auto auto',
          gridGap: theme.spacing(3),
          height: '100vh',
          backgroundImage: 'url(https://gw.alipayobjects.com/zos/rmsportal/TVYTbAXWheQpRcWDaDMu.svg)',
          backgroundRepeat: 'no-repeat',
          backgroundPosition: 'center 110px',
          backgroundSize: '100%'
        },
        margin: {
          margin: '0 15px 0 15px',
        },
        paper: {
          color: theme.palette.text.secondary,
          gridTemplateRows: '1% 30% auto 11% 5% 14% 10%',
          gridColumn: '3/span 1',
          gridRow: '2/span 1',
          display: 'grid',
          gridGap: theme.spacing(1)
        },
        divider: {
          margin: theme.spacing(2, 0),
        },
        backgroundContainer: {
          background: '#f0f2f5'
        }
      }),
    )()
  };
  // Test function to handle react-router and redux connection at the same time
  let historyApp = useHistory();
  const handleClick = () => {
    if (username !== "" && password !== "") {
      props.doLogin(username, password);
      historyApp.push('/home');
    }
    // silently do not submit the form as input color validation is enough to inform the user
  };
  const classes = useStyles();
  return (
    <div className={classes.backgroundContainer}>
      <div className={classes.container}>
        <Paper className={classes.paper}>
          <div style={{ gridRow: '2', justifySelf: 'center', alignSelf: 'center', display: 'grid' }}>
            <img style={{ width: '50px', justifySelf: 'center', alignSelf: 'center' }} src={logo} alt="logo" />
            <Typography variant='h5'>Platter</Typography>
          </div>
          <ValidationTextField
            style={{ gridRow: '3' }}
            className={classes.margin}
            label="Username"
            required
            onChange={(e: any) => setUsername(e.target.value)}
            variant="outlined"
            id="validation-outlined-input"
          />
          <ValidationTextField
            style={{ gridRow: '4' }}
            className={classes.margin}
            label="Password"
            required
            onChange={(e: any) => setPassword(e.target.value)}
            type="password"
            variant="outlined"
            id="validation-outlined-input"
          />
          <Button
            onClick={() => handleClick()}
            style={{ gridRow: '6' }}
            className={classes.margin}
            variant="contained"
            color="primary">
            Login
            </Button>
        </Paper>
      </div>
    </div>
  );
}

export default App;
