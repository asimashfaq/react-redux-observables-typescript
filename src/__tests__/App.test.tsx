import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import App from '../App.connect';
import { createMemoryHistory } from 'history'
import { ConnectedRouter } from 'connected-react-router';
import { Router } from 'react-router-dom';

const mockStore = configureStore([]);
describe('Check render of App Component with store', () => {
    let store;
    let appComponent: any;
    let history;

    beforeEach(() => {
        store = mockStore({
            admin: false,
            status: "",
            token: "",
            name: ""
        });
        history = createMemoryHistory();
        appComponent = renderer.create(
            <Provider store={store} >
                <Router history={history}>
                    <App />
                </Router>
            </Provider>
        );
    });

    it('renders without crashing', () => {
        expect(appComponent.toJSON()).toMatchSnapshot();
    });
})
