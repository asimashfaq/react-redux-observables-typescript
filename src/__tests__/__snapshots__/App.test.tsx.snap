// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check render of App Component with store renders without crashing 1`] = `
<div
  className="makeStyles-backgroundContainer-5"
>
  <div
    className="makeStyles-container-1"
  >
    <div
      className="MuiPaper-root MuiPaper-elevation1 makeStyles-paper-3 MuiPaper-rounded"
    >
      <div
        style={
          Object {
            "alignSelf": "center",
            "display": "grid",
            "gridRow": "2",
            "justifySelf": "center",
          }
        }
      >
        <img
          alt="logo"
          src="test-file-stub"
          style={
            Object {
              "alignSelf": "center",
              "justifySelf": "center",
              "width": "50px",
            }
          }
        />
        <h5
          className="MuiTypography-root MuiTypography-h5"
        >
          Platter
        </h5>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-63 makeStyles-margin-2"
        style={
          Object {
            "gridRow": "3",
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
          htmlFor="validation-outlined-input"
          id="validation-outlined-input-label"
        >
          Username
          <span
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="validation-outlined-input"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-124 MuiOutlinedInput-notchedOutline"
            style={
              Object {
                "paddingLeft": 8,
              }
            }
          >
            <legend
              className="PrivateNotchedOutline-legend-125"
              style={
                Object {
                  "width": 0.01,
                }
              }
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiTextField-root WithStyles(ForwardRef(TextField))-root-63 makeStyles-margin-2"
        style={
          Object {
            "gridRow": "4",
          }
        }
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
          data-shrink={false}
          htmlFor="validation-outlined-input"
          id="validation-outlined-input-label"
        >
          Password
          <span
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoFocus={false}
            className="MuiInputBase-input MuiOutlinedInput-input"
            disabled={false}
            id="validation-outlined-input"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
          />
          <fieldset
            aria-hidden={true}
            className="PrivateNotchedOutline-root-124 MuiOutlinedInput-notchedOutline"
            style={
              Object {
                "paddingLeft": 8,
              }
            }
          >
            <legend
              className="PrivateNotchedOutline-legend-125"
              style={
                Object {
                  "width": 0.01,
                }
              }
            >
              <span
                dangerouslySetInnerHTML={
                  Object {
                    "__html": "&#8203;",
                  }
                }
              />
            </legend>
          </fieldset>
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-margin-2 MuiButton-containedPrimary"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "gridRow": "6",
          }
        }
        tabIndex={0}
        type="button"
      >
        <span
          className="MuiButton-label"
        >
          Login
        </span>
      </button>
    </div>
  </div>
</div>
`;
